{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein motif search and visualization tool\n",
    "Tools to search conserved motif (e.g. CDK phosphorylation) from multiple sequence alignment of different Drosophila species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import library\n",
    "\n",
    "import os\n",
    "import re\n",
    "from glob import glob\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions for different text file format\n",
    "\n",
    "def generate_working_file(input_path):\n",
    "    ''' Read \"original.txt\" file that contains sequences of homologous proteins \n",
    "    in fasta format and generate \"working.txt\" file with removed newlines\n",
    "    Args:\n",
    "        input_path: (str) file path to \"original.txt\"\n",
    "    '''\n",
    "    original_path = os.path.join(input_path, 'original.txt')\n",
    "    species_dict = dict()\n",
    "    with open(original_path, 'r') as f:\n",
    "        for l in f:\n",
    "            if l.startswith('>'):\n",
    "                species = re.findall('\\[(.+)\\]', l)[0] # regular expression to scrape species name within square brackets[]\n",
    "                species_dict[species] = ''\n",
    "            elif len(l.split()) > 0: species_dict[species] += l.split()[0]\n",
    "                \n",
    "    working_path = os.path.join(input_path, 'working.txt')\n",
    "    with open(working_path, 'w') as f:\n",
    "        for s in species_dict.keys():\n",
    "            f.write(('>'+'_'.join(s.split(' '))+'\\n'))\n",
    "            f.write(species_dict[s] + '\\n')\n",
    "\n",
    "\n",
    "def generate_species_dict(input_path):\n",
    "    ''' Read \".clustal_num\" file generated from multiple sequence alignment using clustalw and\n",
    "    generate \"species_dict.txt\" file in fasta format.\n",
    "    Args:\n",
    "        input_path: (str) file path to \".clustal_num\" file\n",
    "    Return:\n",
    "        species_dict: (dict) of format {species: aligned sequence}\n",
    "    '''\n",
    "    msa_fname = [f for f in os.listdir(input_path) if f.endswith('.clustal_num')]\n",
    "    msa_fname = msa_fname[0]\n",
    "    msa_path = os.path.join(input_path, msa_fname)\n",
    "    \n",
    "    species_dict = dict()\n",
    "    with open(msa_path, 'r') as f:\n",
    "        for l in f:\n",
    "            m = re.search('[0-9][0-9]$', l) # line with species name ends with number\n",
    "            if m:\n",
    "                temp_info = l.split()\n",
    "                species = temp_info[0]\n",
    "                if species not in species_dict.keys():\n",
    "                    species_dict[species] = temp_info[1]\n",
    "                else:\n",
    "                    species_dict[species] += temp_info[1]\n",
    "                    \n",
    "    output_path = os.path.join(input_path, \"Analysis\")\n",
    "    if not os.path.exists(output_path):os.mkdir(output_path)\n",
    "\n",
    "    species_path = os.path.join(output_path, 'species_dict.txt')\n",
    "    species_file = open(species_path, 'w')\n",
    "    \n",
    "    for species in species_dict.keys():\n",
    "        species_file.write('>{}\\n'.format(species))\n",
    "        species_file.write(species_dict[species][:-200] + '\\n')\n",
    "    species_file.close()\n",
    "    return species_dict\n",
    "\n",
    "\n",
    "def show_msa_motif(input_path, anchor_species='Drosophila_melanogaster', motifs=['S-*[ST]-*P',]):\n",
    "    ''' Read \".clustal_num\" file generated from multiple sequence alignment and find motif that are\n",
    "    present across different species.\n",
    "    Args:\n",
    "        input_path: (str) file path to \".clustal_num\" file\n",
    "        anchor_species: (str) species of interest to search for motif on\n",
    "        motifs: (list) of (str) regular expression to find specific motifs\n",
    "    '''\n",
    "    species_dict = generate_species_dict(input_path)\n",
    "    anchor_sequence = species_dict[anchor_species]\n",
    "    \n",
    "    max_len = 0\n",
    "    for s in list(species_dict.keys()):\n",
    "        if len(s) > max_len:\n",
    "            max_len = len(s)\n",
    "    max_len = max_len + 6\n",
    "    \n",
    "    output_path = os.path.join(input_path, \"Analysis\")\n",
    "    if not os.path.exists(output_path): os.mkdir(output_path)\n",
    "    \n",
    "    # Generate text file with specific format\n",
    "    for motif in motifs:\n",
    "        pos_start = [m.start(0) for m in re.finditer(motif, anchor_sequence)]\n",
    "        pos_end = [m.end(0) for m in re.finditer(motif, anchor_sequence)]\n",
    "        \n",
    "        save_dict=dict()\n",
    "        for s in list(species_dict.keys()):\n",
    "            temp = []\n",
    "            curr_sequence = species_dict[s]\n",
    "            for i in range(len(pos_start)):\n",
    "                start_sequence = ''.join(re.split('-+',curr_sequence[:pos_start[i]]))\n",
    "                pos_start_str = ' '*(6 - len(str(len(start_sequence)))) + str(len(start_sequence)) + '    '\n",
    "\n",
    "                end_sequence = ''.join(re.split('-+',curr_sequence[:pos_end[i]]))\n",
    "                pos_end_str = ' '*(6 - len(str(len(end_sequence)))) + str(len(end_sequence))\n",
    "\n",
    "                temp.append(pos_start_str+'-'+curr_sequence[pos_start[i]:pos_end[i]]+'-'+pos_end_str)\n",
    "\n",
    "            save_dict[s] = temp\n",
    "        \n",
    "        new_msa_path = os.path.join(output_path, 'msa_motif-'+motif+'.txt')\n",
    "        new_msa_file = open(new_msa_path, 'w')\n",
    "        \n",
    "        for i in range(len(temp)):\n",
    "            for s in list(save_dict.keys()):\n",
    "                new_msa_file.write(s+' '*(max_len - len(s))+save_dict[s][i]+'\\n')\n",
    "            new_msa_file.write('\\n')\n",
    "            new_msa_file.write('\\n')\n",
    "        new_msa_file.close()\n",
    "        \n",
    "\n",
    "def msa_motif_summary(input_path, motifs=['S-*[ST]-*P',]):\n",
    "    ''' Generate \"info.txt\" that summarizes the frequency of different motifs on different species.\n",
    "    Args:\n",
    "        input_path: (str) file path to \".clustal_num\" file\n",
    "        motifs: (list) of (str) regular expression to find specific motifs\n",
    "    '''\n",
    "    species_dict = generate_species_dict(input_path)\n",
    "    \n",
    "    output_path = os.path.join(input_path, \"Analysis\")\n",
    "    if not os.path.exists(output_path):\n",
    "        os.mkdir(output_path)\n",
    "    \n",
    "    general_info_path = os.path.join(output_path, 'info.txt')\n",
    "    general_info_file = open(general_info_path, 'w')\n",
    "    \n",
    "    # Generate text file with specific format\n",
    "    for s in list(species_dict.keys()):\n",
    "        curr_sequence = species_dict[s]\n",
    "        trim_sequence = ''.join(re.split('-+', curr_sequence))\n",
    "        general_info_file.write(s + '\\n')\n",
    "        general_info_file.write('\\n')\n",
    "        for motif in motifs:\n",
    "            general_info_file.write(motif + '\\n')\n",
    "            general_info_file.write('number: ' + str(len(re.findall(motif, curr_sequence))) + '\\n')\n",
    "            general_info_file.write('percentage (no. of motif / total aa): ' + str(len(re.findall(motif, curr_sequence)) / len(trim_sequence) * 100) + '\\n')\n",
    "            general_info_file.write('\\n')\n",
    "        general_info_file.write('=============================\\n')\n",
    "    general_info_file.close()\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions for plotting the location of different motifs\n",
    "\n",
    "def plot_collapsed_msa_motif(input_path, anchor_species='Drosophila_melanogaster', motifs=['S-*[ST]-*P',], styles=['r',]):\n",
    "    ''' Generate a collapsed (narrow) graph of different conserved motifs on a protein.\n",
    "    Args:\n",
    "        input_path: (str) file path to \".clustal_num\" file\n",
    "        motifs: (list) of (str) regular expression to find specific motifs\n",
    "        styles: (list) of (str) color code used by matplotlib.pyplot. It should have the same length as motifs\n",
    "    '''\n",
    "    species_dict = generate_species_dict(input_path)\n",
    "    anchor_sequence = species_dict[anchor_species]\n",
    "    \n",
    "    output_path = os.path.join(input_path, \"Figures\")\n",
    "    if not os.path.exists(output_path):\n",
    "        os.mkdir(output_path)\n",
    "    \n",
    "    motif_dict = dict()\n",
    "    for motif in motifs:\n",
    "\n",
    "        pos_start = [m.start(0) for m in re.finditer(motif, anchor_sequence)]\n",
    "        pos_end = [m.end(0) for m in re.finditer(motif, anchor_sequence)]\n",
    "\n",
    "        tot_count = 0\n",
    "        count = 0\n",
    "\n",
    "        pos_percentage = []\n",
    "\n",
    "        for i in range(len(pos_start)):\n",
    "            tot_count = 0\n",
    "            count = 0\n",
    "            for s in list(species_dict.keys()):\n",
    "                curr_sequence = species_dict[s]\n",
    "                if re.search(motif, curr_sequence[pos_start[i]:pos_end[i]]):\n",
    "                    count += 1\n",
    "                tot_count += 1\n",
    "\n",
    "            start_sequence = ''.join(re.split('-+',anchor_sequence[:pos_start[i]]))\n",
    "            curr_pos = len(start_sequence)\n",
    "            percentage = float(count) / tot_count\n",
    "            pos_percentage.append((curr_pos, percentage))\n",
    "\n",
    "        motif_dict[motif] = pos_percentage\n",
    "        \n",
    "    style_dict = {k: s for k, s in zip(list(motif_dict.keys()), styles)}\n",
    "    trim_sequence = ''.join(re.split('-+', anchor_sequence))\n",
    "    \n",
    "    plt.figure(figsize=(10, 0.5), dpi=150)\n",
    "    for motif in list(motif_dict.keys())[::-1]:\n",
    "        to_label = True\n",
    "        for pos, percentage in motif_dict[motif]:\n",
    "            if percentage >= 1:\n",
    "                if to_label:\n",
    "                    plt.axvline(x=pos, ymin=0, ymax=percentage, color=style_dict[motif], label=motif, alpha=0.9, lw=1)\n",
    "                    to_label = False\n",
    "                else:\n",
    "                    plt.axvline(x=pos, ymin=0, ymax=percentage, color=style_dict[motif], alpha=0.9, lw=1)\n",
    "                    \n",
    "    plt.axvline(x=0, ymin=0, ymax=0, color='white', alpha=0, label='                                 ')\n",
    "    plt.xlim((0, len(trim_sequence)))\n",
    "    plt.ylim((0, 1))\n",
    "    plt.yticks([])\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(os.path.join(output_path, 'collapsed_motif-{0}.png'.format(list(motif_dict.keys())[::-1])), bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    \n",
    "    \n",
    "def plot_motif_domain(input_path, anchor_species='Drosophila_melanogaster', motifs=['S-*[ST]-*P',], styles=['r',], filter_percentage=0.):\n",
    "    ''' Generate a graph showning the degree of conservation of different motifs on a protein from certain species.\n",
    "    Args:\n",
    "        input_path: (str) file path to \".clustal_num\" file\n",
    "        anchor_species: (str) species of interest to search for motif on\n",
    "        motifs: (list) of (str) regular expression to find specific motifs\n",
    "        filter_percentage: (float) at least this percentage of conservation to be shown on the graph\n",
    "    '''\n",
    "    output_path = os.path.join(input_path, \"Figures\")\n",
    "    if not os.path.exists(output_path):\n",
    "        os.mkdir(output_path)\n",
    "    \n",
    "    species_dict = generate_species_dict(input_path)\n",
    "    anchor_sequence = species_dict[anchor_species]\n",
    "    motif_dict = dict()\n",
    "    for motif in motifs:\n",
    "        pos_start = [m.start(0) for m in re.finditer(motif, anchor_sequence)]\n",
    "        pos_end = [m.end(0) for m in re.finditer(motif, anchor_sequence)]\n",
    "\n",
    "        tot_count = 0\n",
    "        count = 0\n",
    "        pos_percentage = []\n",
    "        for i in range(len(pos_start)):\n",
    "            tot_count = 0\n",
    "            count = 0\n",
    "            for s in list(species_dict.keys()):\n",
    "                curr_sequence = species_dict[s]\n",
    "                if re.search(motif, curr_sequence[pos_start[i]:pos_end[i]]):\n",
    "                    count += 1\n",
    "                tot_count += 1\n",
    "            start_sequence = ''.join(re.split('-+',anchor_sequence[:pos_start[i]]))\n",
    "            curr_pos = len(start_sequence)\n",
    "            percentage = float(count) / tot_count\n",
    "            pos_percentage.append((curr_pos, percentage))\n",
    "        motif_dict[motif] = pos_percentage\n",
    "        \n",
    "    style_dict = {k: s for k, s in zip(list(motif_dict.keys()), styles)}\n",
    "    trim_sequence = ''.join(re.split('-+', anchor_sequence))\n",
    "\n",
    "    plt.figure(figsize=(10, 2.5), dpi=150)\n",
    "    \n",
    "    for motif in list(motif_dict.keys())[::-1]:\n",
    "        to_label = True\n",
    "\n",
    "        for pos, percentage in motif_dict[motif]:\n",
    "            if percentage >= filter_percentage:\n",
    "                if to_label:\n",
    "                    plt.axvline(x=pos, ymin=0, ymax=percentage, color=style_dict[motif], label=motif, alpha=0.9, lw=0.5)\n",
    "                    to_label = False\n",
    "                else:\n",
    "                    plt.axvline(x=pos, ymin=0, ymax=percentage, color=style_dict[motif], alpha=0.9, lw=0.5)\n",
    "    \n",
    "    plt.axvline(x=0, ymin=0, ymax=0, color='white', alpha=0, label='                                 ')\n",
    "    plt.legend(bbox_to_anchor=(1.0, 1), loc='upper left')\n",
    "    plt.xlabel('amino acid')\n",
    "    plt.ylabel('fraction')\n",
    "    plt.xlim((0, len(trim_sequence)))\n",
    "    plt.ylim((0, 1))\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(os.path.join(output_path, 'plot_filter-{0}_motif-{1}.png'.format(filter_percentage, len(list(motif_dict.keys())))), bbox_inches=\"tight\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis for a single protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some parameters to be filled\n",
    "input_path = 'xxxxxx/xxxxx/xxxxxx'\n",
    "anchor_species = 'Drosophila_melanogaster'\n",
    "motifs = ['[ST]-*P',]\n",
    "styles = ['r'] # same length as motifs\n",
    "degree_of_conservation = 0.5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate \"working.txt\" from \"original.txt\" which you copy and paste homologous proteins from different species\n",
    "generate_working_file(input_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gereate text file to show the position of different motifs\n",
    "show_msa_motif(input_path, anchor_species, motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summarize the frequency of different motifs on homologous proteins from different species\n",
    "msa_motif_summary(input_path, motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the graph of conservation\n",
    "plot_motif_domain(input_path,  anchor_species, motifs, styles, degree_of_conservation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:centrosome] *",
   "language": "python",
   "name": "conda-env-centrosome-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
